### ISSUE TITLE
New Lint: Use question mark instead of Option::and_then

### ISSUE THREAD

**Issue by camsteffen** [rust-lang/rust-clippy#6436]
### What it does

Detect functions that end with `Option::and_then` or `Result::and_then` and suggest using a question mark instead.

### Categories

- Kind: style, maybe pedantic

It is simpler and more idiomatic.

### Drawbacks

None.

### Example

```rust
fn test(opt: Option<i32>) -> Option<i32> {
    opt.and_then(|n| {
        if n > 1 {
            Some(n + 1)
        } else {
            None
        }
    })
}
```

Could be written as:

```rust
fn test(opt: Option<i32>) -> Option<i32> {
    let n = opt?;
    if n > 1 {
        Some(n + 1)
    } else {
        None
    }
}
```

The `and_then` call should be the very last expression of the function (not inside an `if` block, for example).

### More Questionable Cases

Single expression lambda (no block)

```rust
fn test(opt: Option<i32>) -> Option<i32> {
    opt.and_then(|n| foo(n))
}
```

At the end of a call chain

```rust
fn foo() -> Option<i32> {
    vec![1, 2, 3]
       .iter()
       .last()
       .and_then(|n| {
           // ...
       })
}
```

Personally I think those cases should still be linted. Maybe it can be configurable.

**Comment by hristotanev** [comment]
just checking if this is still good to be worked on? i am aware it has been a while! if it is, i don't mind picking it up :)!

**Comment by Manishearth** [comment]
Should be fine. We should figure out the name and the lint level but you can work on it anyway.

I'm open to the level being `style` or `complexity` as long as lintcheck isn't too noisy

**Comment by sbdchd** [comment]
I found myself rewriting:

```rust
let select = paren_select(p);
match select {
    Some(cm) => {
        if p.at_ts(COMPOUND_SELECT_FIRST) {
            opt_compound_select(p, cm)
        } else {
            Some(cm)
        }
    }
    None => None,
}
```

as

```rust
paren_select(p).and_then(|cm| {
    if p.at_ts(COMPOUND_SELECT_FIRST) {
        opt_compound_select(p, cm)
    } else {
        Some(cm)
    }
})
```

before switching to the question mark:

```rust
let cm = paren_select(p)?;
if p.at_ts(COMPOUND_SELECT_FIRST) {
    opt_compound_select(p, cm)
} else {
    Some(cm)
}
```
